src/dir.rs

```diff
@@ -0,0 +1,208 @@
+use {Error, NixPath, Result};
+use errno::Errno;
+use fcntl::{self, OFlag};
+use libc;
+use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
+use std::{ffi, fmt, ptr};
+use sys;
+
+#[cfg(target_os = "linux")]
+use libc::{dirent64 as dirent, readdir64_r as readdir_r};
+
+#[cfg(not(target_os = "linux"))]
+use libc::{dirent, readdir_r};
+
+/// An open directory.
+///
+/// This is a lower-level interface than `std::fs::ReadDir`. Notable differences:
+///    * can be opened from a file descriptor (as returned by `openat`, perhaps before knowing
+///      if the path represents a file or directory).
+///    * implements `AsRawFd`, so it can be passed to `fstat`, `openat`, etc.
+///      The descriptor continues to be owned by the `Dir`, so callers must not keep a `RawFd`
```

Comment:
s/descriptor/file/

Reply:
file descriptor? it's really not the file.

Reply:
Hmm, I think this argument applies equally well to anywhere else that uses `AsRawFd`, and Google doesn't find any results for `AsRef<RawFd>`. I think I'd prefer to match the rest of the ecosystem.

Reply:
and `RawFd` is `Copy`, so `AsRef<RawFd>` is probably just as easy to mess up anyway.

Reply:
I forgot about `Copy`.  In that case you're right that `AsRef<RawRd>` wouldn't have much advantage.